### PR TITLE
Add apiextensions.CustomResource to Go SDK

### LINKED ISSUE #1116: [Go SDK] Add apiextensions.CustomResource
### Problem description

Add the `apiextensions.CustomResource` class to the Go SDK.

### PR DESCRIPTION
**Pull request by lblackstone**
### Proposed changes
This change makes it possible to create k8s `CustomResource` objects from a YAML manifest. Still TODO:

- [x] Finish Output types - waiting on https://github.com/pulumi/pulumi/issues/4629

### Related issues (optional)
Fixes #1116

### CODE REVIEW

File: provider/pkg/gen/go-templates/apiextensions/customResource.tmpl [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+type CustomResourceArgs struct {
+	// APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources
+	ApiVersion pulumi.StringInput
+	// Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	Kind pulumi.StringInput
+	// Standard object metadata.
+	Metadata metav1.ObjectMetaPtrInput
+	// Untyped map that holds any user-defined fields.
+	// TODO: do we need to support Inputs in the map?
+	OtherFields kubernetes.UntypedArgs
```
Comment (lukehoban) [inline provider/pkg/gen/go-templates/apiextensions/customResource.tmpl:126]: Two terminology questions (I don't feel strongly - but wanted to ask):
* Are these "fields" or "properties" in the Kubernetes object model?
* Is "Other" needed?

File: provider/pkg/gen/go-templates/apiextensions/customResource.tmpl [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+func NewCustomResource(ctx *pulumi.Context,
+	name string, args *CustomResourceArgs, opts ...pulumi.ResourceOption) (*CustomResource, error) {
+	if args == nil {
+		args = &CustomResourceArgs{}
+	}
+
+	untyped := args.OtherFields
+	untyped["apiVersion"] = args.ApiVersion
+	untyped["kind"] = args.Kind
+	untyped["metadata"] = args.Metadata
```
Comment (lukehoban) [inline provider/pkg/gen/go-templates/apiextensions/customResource.tmpl:60]: Will this mutate the dict passed in by the user?  Should we copy fields into a new dict here?

File: provider/pkg/gen/go-templates/apiextensions/customResource.tmpl [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+// The set of arguments for constructing a Deployment resource.
+type CustomResourceArgs struct {
+	// APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources
+	ApiVersion pulumi.StringInput
+	// Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	Kind pulumi.StringInput
+	// Standard object metadata.
+	Metadata metav1.ObjectMetaPtrInput
+	// Untyped map that holds any user-defined fields.
+	// TODO: do we need to support Inputs in the map?
```
Comment (lukehoban) [inline provider/pkg/gen/go-templates/apiextensions/customResource.tmpl:125]: I suspect we do - and I think that is possible with the `UntypedArgs` support perhaps?

File: sdk/go/kubernetes/apps/v1beta1/controllerRevision.go [hunk 4]
```diff
@@ -12,6 +12,8 @@ import (
 )
 
 // ControllerRevision implements an immutable snapshot of state data. Clients are responsible for serializing and deserializing the objects that contain their internal state. Once a ControllerRevision has been successfully created, it can not be updated. The API Server will fail validation of all requests that attempt to mutate the Data field. ControllerRevisions may, however, be deleted. Note that, due to its use by both the DaemonSet and StatefulSet controllers for update and rollback, this object is beta. However, it may be subject to name and representation changes in future releases, and clients should not depend on its stability. It is primarily for internal use by controllers.
+//
+// Deprecated: apps/v1beta1/ControllerRevision is deprecated by apps/v1/ControllerRevision and not supported by Kubernetes v1.16+ clusters.
```
Comment (lukehoban) [inline sdk/go/kubernetes/apps/v1beta1/controllerRevision.go:16]: Are these changes intentional?

File: tests/integration/go/yaml/manifests/crontab.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (lukehoban) [inline tests/integration/go/yaml/manifests/crontab.yaml:1]: How are we automating testing this if the CRD needs to be installed?

Should we install this (or a simpler) CRD as part of the Go program?

File: tests/integration/go/yaml/main.go [hunk 6]
```diff
@@ -43,6 +44,14 @@ func main() {
...
 		ctx.Export("hostIP", hostIP)
 
+		ct := resources.GetResource("stable.example.com/v1/CronTab", "my-new-cron-object", "")
+		cronSpec := ct.(*apiextensions.CustomResource).OtherFields.ApplyT(func(otherFields interface{}) string {
+			fields := otherFields.(map[string]interface{})
+			spec := fields["spec"].(map[string]interface{})
+			return spec["cronSpec"].(string)
+		})
+		ctx.Export("cronSpec", cronSpec)
+
```
Comment (lukehoban) [inline tests/integration/go/yaml/main.go:54]: Am I right that there is not yet a test covering `NewCustomResource`?  That seems worth adding.

File: tests/integration/go/yaml/manifests/crontab.yaml (hunk 5 above)
Comment (lblackstone) [reply tests/integration/go/yaml/manifests/crontab.yaml:1]: The test will install both the CRD and CR from this manifest. The built in retry allows it to succeed even though they’re in the same file.

File: provider/pkg/gen/go-templates/apiextensions/customResource.tmpl [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+)
+
+// CustomResource represents a resource definition we'd use to create an instance of a
+// Kubernetes CustomResourceDefinition (CRD). For example, the CoreOS Prometheus operator
+// exposes a CRD `monitoring.coreos.com/ServiceMonitor`; to create a `ServiceMonitor`, we'd
+// pass a `CustomResourceArgs` containing the `ServiceMonitor` definition to
+// `apiextensions.CustomResource`.
+//
+// NOTE: This type is abstract. You need to inherit from it and define all the properties of
+// a specific custom resource in the derived class.
```
[the remark this answers is not in the document]
Comment (lblackstone) [reply provider/pkg/gen/go-templates/apiextensions/customResource.tmpl:36]: Hmm, probably not. I copied that from the C# implementation.

File: provider/pkg/gen/go-templates/apiextensions/customResource.tmpl (hunk 2 above)
Comment (lblackstone) [reply provider/pkg/gen/go-templates/apiextensions/customResource.tmpl:60]: Possibly so. I’ll copy to a new map

File: provider/pkg/gen/go-templates/apiextensions/customResource.tmpl (hunk 3 above)
Comment (lblackstone) [reply provider/pkg/gen/go-templates/apiextensions/customResource.tmpl:125]: I’ll test that

File: sdk/go/kubernetes/apps/v1beta1/controllerRevision.go (hunk 4 above)
Comment (lblackstone) [reply sdk/go/kubernetes/apps/v1beta1/controllerRevision.go:16]: Yeah, they got pulled in due to codegen changes in p/p

File: provider/pkg/gen/go-templates/apiextensions/customResource.tmpl (hunk 1 above)
Comment (lblackstone) [reply provider/pkg/gen/go-templates/apiextensions/customResource.tmpl:126]: The API reference calls them "fields". I copied the name from the TypeScript SDK, but don't have a strong opinion on it.

File: provider/pkg/gen/go-templates/apiextensions/customResource.tmpl (hunk 3 above)
Comment (lblackstone) [reply provider/pkg/gen/go-templates/apiextensions/customResource.tmpl:125]: Yep, it works as expected.

File: tests/integration/go/yaml/manifests/crontab.yaml (hunk 5 above)
Comment (lukehoban) [reply tests/integration/go/yaml/manifests/crontab.yaml:1]: I can't tell where this file is actually getting used - are we running a test that uses this?

Comment (lblackstone) [reply tests/integration/go/yaml/manifests/crontab.yaml:1]: Yeah, it gets applied by the `ConfigGroup` in the yaml test.

**Comment by lblackstone** [comment]
> Have we manually tried a non-trivial Helm chart (or group of Config) to validate that it works well in the full E2E?

I just tried a bunch of Helm charts, and they all worked fine. I think the existing test coverage is exercising all of the basics now (YAML file, YAML folder, Helm Chart, CRDs + CRs).
